### PR TITLE
Implement Sigv4a Signing in aws_request_signing_extension

### PR DESCRIPTION
**Pull request by nbaws**
Commit Message: Implement sigv4a signing in aws_request_signing extension
Additional Description: See https://github.com/envoyproxy/envoy/issues/31000 for proposal
Risk Level: Low
Testing: Unit
Docs Changes: API Proto
Release Notes: N/A
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/31000
[Optional Fixes commit #PR or SHA] 
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/crypto/utility.h [hunk 1]
```diff
@@ -47,6 +47,15 @@ class Utility {
   virtual std::vector<uint8_t> getSha256Hmac(const std::vector<uint8_t>& key,
                                              absl::string_view message) PURE;
 
+  /**
+   * Computes the SHA-256 HMAC for a given key and message.
+   * @param key the HMAC function key.
+   * @param message message data for the HMAC function.
```
[the remark this answers is not in the document]
Comment (suniltheta) [reply source/common/crypto/utility.h:53]: Meanwhile is there an option to re-use the existing function itself?

So that no code change is required on this common code as we as unit test update

File: api/envoy/config/route/v3/route_components.proto [hunk 2]
```diff
@@ -2365,6 +2365,7 @@ message QueryParameterMatcher {
 }
 
 // HTTP Internal Redirect :ref:`architecture overview <arch_overview_internal_redirects>`.
+// [#next-free-field: 6]
```
Comment (suniltheta) [inline api/envoy/config/route/v3/route_components.proto:2368]: unrelated change.

**Review by suniltheta — COMMENTED** [review]
Feel free to keep the issue #31000 still open and mentioned on the description like below so that the Issue is auto closed when the PR is merged

```
Fixes: https://github.com/envoyproxy/envoy/issues/31000
```

File: source/extensions/common/aws/sigv4a_signer_impl.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+using SigV4ASignatureConstants = ConstSingleton<SigV4ASignatureConstantValues>;
+
+using AwsSigningHeaderExclusionVector = std::vector<envoy::type::matcher::v3::StringMatcher>;
+
+/**
+ * Implementation of the Signature V4A signing process.
+ * See https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+ */
+
+class SigV4ASignerImpl : public Signer, public Logger::Loggable<Logger::Id::aws> {
```
Comment (suniltheta) [inline source/extensions/common/aws/sigv4a_signer_impl.h:60]: is it fair to say some of the functions are the same between the SigV4SignerImpl & SigV4aSignerImpl?

If so can we create a parent class to hold all the common functionalities so that we don't accidentally introduce a bug in future with less repeated code?

File: source/extensions/common/aws/sigv4a_signer_impl.cc [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+  auto& crypto_util = Envoy::Common::Crypto::UtilitySingleton::get();
+
+  EC_KEY* ec_key = SigV4AKeyDerivation::derivePrivateKey(access_key_id, secret_access_key);
+  if (!ec_key) {
+    ENVOY_LOG(debug, "SigV4A key derivation failed");
+    return BlankStr;
+  }
+
+  // // AWS SigV4A Key Derivation
```
Comment (suniltheta) [inline source/extensions/common/aws/sigv4a_signer_impl.cc:143]: I believe these commented code will be removed before the merge?

File: test/extensions/common/aws/sigv4a_signer_impl_test.cc [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+  SigV4AKeyDerivation::derivePublicKey(ec_key);
+
+  Http::TestRequestHeaderMapImpl headers{};
+
+  headers.setMethod("GET");
+  headers.setPath("/");
+  headers.addCopy(Http::LowerCaseString("host"), "www.example.com");
+
+  // Sign the message using our signing algorithm
+  signer_.signUnsignedPayload(headers, "ap-southeast-2,*");
```
Comment (suniltheta) [inline test/extensions/common/aws/sigv4a_signer_impl_test.cc:402]: what is different between `ap-southeast-2,*` and just `*`, wouldn't wildcard `*` cover `ap-southeast-2` as well?

File: source/extensions/common/aws/sigv4a_signer_impl.h (hunk 3 above)
Comment (nbaws) [reply source/extensions/common/aws/sigv4a_signer_impl.h:60]: this is a fair comment. i will look to update this to incorporate common code across both
